File: services/shared/dotnet/Shared.Domain/ValueObjects/Address.cs
namespace Shared.Domain.ValueObjects
{
    public sealed class Address : ValueObject
    {
        public string Street { get; private set; }
        public string City { get; private set; }
        public string State { get; private set; }
        public string PostalCode { get; private set; }
        public string Country { get; private set; }

        public Address(string street, string city, string state, string postalCode, string country)
        {
            if (string.IsNullOrWhiteSpace(street))
                throw new ArgumentException("Street cannot be null or empty");
            if (string.IsNullOrWhiteSpace(city))
                throw new ArgumentException("City cannot be null or empty");
            if (string.IsNullOrWhiteSpace(country))
                throw new ArgumentException("Country cannot be null or empty");

            Street = street;
            City = city;
            State = state ?? string.Empty;
            PostalCode = postalCode ?? string.Empty;
            Country = country;
        }

        protected override IEnumerable<object> GetEqualityComponents()
        {
            yield return Street;
            yield return City;
            yield return State;
            yield return PostalCode;
            yield return Country;
        }

        public override string ToString() => $"{Street}, {City}, {State} {PostalCode}, {Country}";
    }
}

File: client/nextjs/shop-ddd/src/features/orders/domain/value-objects/Address.ts
export class Address {
    private constructor(
        public readonly street: string,
        public readonly city: string,
        public readonly postalCode: string,
        public readonly country: string
    ) {}

    static create(p: { street: string; city: string; postalCode: string; country: string }) {
        for (const [k, v] of Object.entries(p)) {
            if (!v || String(v).trim().length === 0) throw new Error(`Address.${k} required`);
        }
        return new Address(p.street.trim(), p.city.trim(), p.postalCode.trim(), p.country.trim());
    }
}
